src/librustc_target/abi/mod.rs

```diff
@@ -1081,4 +1119,57 @@ impl<'a, Ty> TyLayout<'a, Ty> {
             Abi::Aggregate { sized } => sized && self.size.bytes() == 0
         }
     }
+
+    /// Determines if zero-initializing this type might be okay.
+    /// This is conservative: in doubt, it will answer `true`.
+    pub fn might_permit_zero_init<C, E>(
+        &self,
+        cx: &C,
+    ) -> Result<bool, E>
+    where
+        Self: Copy,
+        Ty: TyLayoutMethods<'a, C>,
+        C: LayoutOf<Ty = Ty, TyLayout: MaybeResult<Self, Error = E>>
+    {
+        fn scalar_allows_zero(s: &Scalar) -> bool {
+            (*s.valid_range.start() <= 0) || // `&& *s.valid_range.end() >= 0` would be redundant
+            (*s.valid_range.start() > *s.valid_range.end()) // wrap-around allows 0
+        }
+
+        // Abi is the most informative here.
+        let res = match &self.abi {
+            Abi::Uninhabited => false, // definitely UB
+            Abi::Scalar(s) => scalar_allows_zero(s),
+            Abi::ScalarPair(s1, s2) =>
+                scalar_allows_zero(s1) && scalar_allows_zero(s2),
+            Abi::Vector { element: s, count } =>
+                *count == 0 || scalar_allows_zero(s),
+            Abi::Aggregate { .. } => {
+                // For aggregates, recurse.
+                let inner = match self.variants {
+                    Variants::Multiple { .. } => // FIXME: get variant with "0" discriminant.
+                        return Ok(true),
+                    Variants::Single { index } => self.for_variant(&cx, index),
```

Comment:
`for_variant` is a noop for `Variant::Single`.

Reply:
Is that a stable guarantee? Miri's value visitor contains similar code. I think it makes the structure quite clear.

Reply:
Yeah, it turns `Multiple` into `Single`, the only reason to apply it in the `Single` case is if you have a variant index and you *don't* want to pattern-match on `.variants`.

Reply:
Okay. I was also wrong about the value visitor, it actually doesn't call `for_variant` for `Single`. So I adjusted the code here as well.